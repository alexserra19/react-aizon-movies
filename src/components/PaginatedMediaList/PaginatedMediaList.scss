@import '../../styles/variables.scss';

.paginatedMediaListContainer {
    .noMediaAvailable{
        text-align: center;
        padding: 50px;
        h4 {
            color: $white;
        }
    }

    .gridContainer {
        padding: 0;
    }

    padding: 20px 0px;

    h3 {
        font-weight: bold;
        color: $white;
    }

    .mediaItemContainer {
        padding: 5px;
        .mediaItemBody {
            position: relative;
            height: 250px;
    
            img {
                height: 100%;
                width: 100%;
            }
    
            .titleContainer, .titleContainerDescription {
                position: absolute;
                bottom: 0;
                width: 100%;
                background-color: $black;
                opacity: 0.8;
                padding: 15px;

                span {
                    color: $white
                }
            }

            .titleContainerDescription {
                display: none;
                .bx--grid {
                    padding: 5px 15px;
                }

                b {
                    font-weight: bold;
                    color: $white;
                }

                .year {
                    color: $red
                }

                span {
                    margin: 2px 0px;
                }
            }

        }
        a:hover {
            .titleContainer {
                display: none;
            }
            .titleContainerDescription {
                display: inherit;
            }
        }
    }

    .paginator {
        text-align: center;
        margin: 40px 0px;

        .ant-pagination-item-link {
            background-color: transparent;
            color: white;
            border: 0;
        }

        .ant-pagination-item {
            color: $white;
            background-color: $red;
            border: none;
            border-radius: 20px;
            opacity: 0.5;
            a {
                color: $white;
            }
        }

        .ant-pagination-item-active {
            opacity: 1;
        }
    }
    
    .bx--loading__stroke {
        stroke: $red;
    }

}