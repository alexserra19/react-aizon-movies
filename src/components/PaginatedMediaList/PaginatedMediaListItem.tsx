import React from 'react';

interface IPaginatedMediaListItemProps {
    title: string;
    image: any;
    date: string;
    description: string;
}

export const PaginatedMediaListItem = (props: IPaginatedMediaListItemProps) => {
    return (
        <div className="bx--col-lg-3 mediaItemContainer" style={{ position: 'relative' }}>
            <a href="#">
                <div className="mediaItemBody">
                    <img src={props.image} className="menuImage" alt={props.title} />
                    <div className="titleContainer">
                        <span>{props.title}</span>
                    </div>
                    <div className="titleContainerDescription">
                        <div className="bx--grid">
                            <div className="bx--row">
                                <b>{props.title}</b>
                            </div>
                            <div className="bx--row">
                                <span className="year">{props.date}</span>
                            </div>
                            <div className="bx--row">
                                <span>{props.description}</span>
                            </div>
                        </div>
                    </div>
                </div>
            </a>
        </div>
    );
}
