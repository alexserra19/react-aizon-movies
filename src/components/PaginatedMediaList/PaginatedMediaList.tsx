
import React, { useEffect, useState } from 'react';
import './PaginatedMediaList.scss'
import { Media } from '../../utils/typings';
import _ from 'lodash';
import { Pagination } from 'antd';
import { Loading } from "carbon-components-react"
import { PaginatedMediaListItem } from './PaginatedMediaListItem';

interface IPaginatedMediaListProps {
    title: string;
    mediaList: Array<Media>;
    isLoading: boolean
}

export const PaginatedMediaList = (props: IPaginatedMediaListProps) => {

    const [currentPage, setCurrentPage] = useState<number>(1);
    const [paginatedArray, setPaginatedArray] = useState<Array<Media>>(props.mediaList);

    useEffect(() => {
        const length = props.mediaList.length > 10 ? 10 : props.mediaList.length
        const from = (currentPage - 1) * 10
        const to = from + length
        let newArray = []

        for (let index = from; index < to && props.mediaList[index]; index++) {
            newArray.push(props.mediaList[index])
        }

        setPaginatedArray(newArray);

    }, [currentPage, props.mediaList])

    useEffect(() => {
        setCurrentPage(1);
    }, [props.mediaList])


    const renderList = () => {
        return (
            <>
                <div className="bx--grid gridContainer">
                    <div className="bx--row">
                        {paginatedArray.map((media: Media, index: number) => renderItem(media, index))}
                    </div>
                </div>
                <Pagination
                    className="paginator"
                    current={currentPage}
                    onChange={setCurrentPage}
                    defaultPageSize={10}
                    total={props.mediaList.length}
                    showSizeChanger={false}
                />
            </>
        )
    }

    const renderItem = (item: Media, index: number) => {
        return (
            <PaginatedMediaListItem
                title={item.title}
                image={item.image}
                description={item.description}
                date={item.date}
                key={index}
            />
        )
    }

    const renderEmptyListMessage = () => {
        return (
            <div className="noMediaAvailable">
                <h4>No results found</h4>
            </div>
        )
    }

    return (
        <div className="paginatedMediaListContainer">
            <h3>{props.title}</h3>
            {!_.isEmpty(props.mediaList) && renderList()}
            {_.isEmpty(props.mediaList) && !props.isLoading && renderEmptyListMessage()}
            {props.isLoading &&
                <Loading
                    description="Active loading indicator"
                    withOverlay={true}
                    active={true}
                />
            }
        </div>
    );
}
