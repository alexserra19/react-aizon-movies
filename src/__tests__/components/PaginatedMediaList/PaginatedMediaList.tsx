import React from 'react';
import { PaginatedMediaList } from '../../../components/PaginatedMediaList/PaginatedMediaList';
import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';

jest.useFakeTimers()

describe('PaginatedMediaList', () => {

  const mockProps = {
    title: 'title',
    mediaList: [],
    isLoading: true
  }

  it('should match the snapshot', () => {
    const wrapper = render(
      <Router>
        <PaginatedMediaList  {...mockProps} />
      </Router>
    );
    expect(wrapper).toMatchSnapshot();
  })

  it('should render without the loading wheel', () => {

    const props = {
      ...mockProps,
      isLoading: false
    }
    const wrapper = render(
      <Router>
        <PaginatedMediaList  {...props} />
      </Router>
    );

    expect(wrapper.queryByText('Active loading indicator')).toBeNull();
  })
});
