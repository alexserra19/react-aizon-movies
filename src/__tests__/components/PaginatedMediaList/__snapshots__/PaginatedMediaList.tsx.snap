// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginatedMediaList should match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="paginatedMediaListContainer"
      >
        <h3>
          title
        </h3>
        <div
          class="bx--loading-overlay"
        >
          <div
            aria-atomic="true"
            aria-labelledby="loading-id-1"
            aria-live="assertive"
            class="bx--loading"
          >
            <label
              class="bx--visually-hidden"
              id="loading-id-1"
            >
              Active loading indicator
            </label>
            <svg
              class="bx--loading__svg"
              viewBox="0 0 100 100"
            >
              <title>
                Active loading indicator
              </title>
              <circle
                class="bx--loading__stroke"
                cx="50%"
                cy="50%"
                r="44"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="paginatedMediaListContainer"
    >
      <h3>
        title
      </h3>
      <div
        class="bx--loading-overlay"
      >
        <div
          aria-atomic="true"
          aria-labelledby="loading-id-1"
          aria-live="assertive"
          class="bx--loading"
        >
          <label
            class="bx--visually-hidden"
            id="loading-id-1"
          >
            Active loading indicator
          </label>
          <svg
            class="bx--loading__svg"
            viewBox="0 0 100 100"
          >
            <title>
              Active loading indicator
            </title>
            <circle
              class="bx--loading__stroke"
              cx="50%"
              cy="50%"
              r="44"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
